File: assets/js/message.js
export class Message {
  constructor() {}

  setSDP(sdp) {
    if (sdp == null) return;
    this.type = "sdp";
    this.sdp = sdp;
  }

  setICE(ice) {
    if (ice == null) return;
    this.type = "ice";
    this.ice = ice;
  }

  getJSON() {
    return {
      type: this.type,
      action: "publish",
      topic: "webrtc",
      ice: this.ice,
      sdp: this.sdp,
    };
  }
}

File: internal/infrastructure/websocket/message.go
package websocket

import pion "github.com/pion/webrtc/v4"

type Action string

const (
	Publish     Action = "publish"
	Subscribe          = "subscribe"
	Unsubscribe        = "unsubscribe"
)

type MessageType string

const (
	MessageTypeSDP          MessageType = "sdp"
	MessageTypeIceCandidate             = "ice"
	MessageTypeError                    = "error"
)

type TopicId string

const (
	TopicWebRTC TopicId = "webrtc"
)

type Message struct {
	Type   MessageType            `json:"type"`
	Action Action                 `json:"action"`
	Topic  TopicId                `json:"topic"`
	Sdp    string                 `json:"sdp"`
	Ice    *pion.ICECandidateInit `json:"ice"`
	Error  string                 `json:"error"`
}
